images/handlers.go

```diff
@@ -179,14 +181,104 @@ func Dispatch(ctx context.Context, handler Handler, limiter *semaphore.Weighted,
 			if len(children) > 0 {
 				return Dispatch(ctx2, handler, limiter, children...)
 			}
-
 			return nil
 		})
 	}
+	return eg.Wait()
+}
 
+// DispatchWithOpts runs the provided handler for content specified by the
+// descriptors with concurrent and duplication suppressor control.
+//
+// If the handler decodes subresources, they will be visited, as well. A handler
+// can return 'ErrSkipDesc` to signal to the dispatcher to not traverse any
+// children. If any handler returns an error, the dispatch session will be
+// canceled.
+func DispatchWithOpts(ctx context.Context, handler Handler, opts []DispatchOpt, descs ...ocispec.Descriptor) error {
+	control := &dispatchControl{
+		concurrencyLimiter:    pkgsemaphore.NewNoop(),
+		duplicationSuppressor: pkgsingleflight.NewNoop(),
+	}
+
+	for _, opt := range opts {
+		if err := opt(control); err != nil {
+			return err
+		}
+	}
+	return dispatchWithControl(ctx, handler, control, descs...)
+}
+
+func dispatchWithControl(ctx context.Context, handler Handler, control *dispatchControl, descs ...ocispec.Descriptor) error {
+	eg, ctx2 := errgroup.WithContext(ctx)
+	for _, desc := range descs {
+		desc := desc
+
+		if err := control.concurrencyLimiter.Acquire(ctx, 1); err != nil {
+			return err
+		}
+
+		eg.Go(func() error {
+			desc := desc
+
+			if err := control.duplicationSuppressor.Acquire(ctx2, desc.Digest.String()); err != nil {
```

Comment:
good question! I was trying to use this package before. But it will make logic complicated. 

Let say there are several requests on the same layer. First request is in-flight, the rest of them will wait for it. Since the result is shared, the rest of them need to retry if the first one fails(for instance, invalid password) or is canceled. The Dispatch needs to check result and does retry it necessary. It is painful when I use it. Acquire/Release looks easy and flexible to me. But open to suggestion about how to use sync/singleflight package🙏

And I would like to use interface to prevent from nil-checker conditions.

Reply:
Where are such retries taking place? Isn't that way down in the resolver?
The calling context for using x/sync/singleflight should not really change.
I would call `sfGroup.Do(func() (interface{}, error) { return handler.Handle(ctx2, desc) })`.

Really just bringing this up so we aren't implementing/maintaining new code that may not be needed.

Reply:
Well. The `DeplicationSuppressor` is used in several pulling image requests. 

For instance, 

```go
# pulling request 1(in-flght)
err := sfGroup.Do(key, fn-1) with ctx-1

# pulling request 2(in-wait-list)
err := sfGroup.Do(key, fn-2) with ctx-2
```

pulling-request-1's ctx has been canceled and sfGroup.Do returns error like ErrCanceled.
And then sfGroup.Do will wake up pulling-request-2 with the same error, ErrCanceled.
Since there are different context, IMO, pulling-request-2 has no idea how to retry with such kind of error.
Because pulling-request-2's fn-2 doesn't run. 

I think the package sync-sf is used to run the same fn as input. In CRI, the fn has different context.
That is why I choose to maintain new code here.

Reply:
According to @kzys's suggestion, I used kmutex name back and hopefully it is less confusing here.